### PR TITLE
Unify GreeterWidget and SessionWidget

### PR DESCRIPTION
**Pull request by danirabbit**
Keep things from diverging as we go by combining these two classes

It seems like we're getting a crash in the greeter because `org.gnome.desktop.interface` is not available in the greeter session

### CODE REVIEW

**Comment by ryonakano** [comment]
> It seems like we're getting a crash in the greeter because org.gnome.desktop.interface is not available in the greeter session

@danrabbit How did you face a crash? I don't see any problems with this branch and everything looks work as expected.

File: src/SessionWidget.vala [hunk 1]
```diff
@@ -107,6 +140,82 @@ public class A11Y.SessionWidget : Gtk.Grid {
...
+            string[] argv;
+            Shell.parse_argv ("onboard --xid", out argv);
+            Process.spawn_async_with_pipes (null, argv, null, SpawnFlags.SEARCH_PATH, null, out keyboard_pid, null, out onboard_stdout_fd, null);
+
+            var f = FileStream.fdopen (onboard_stdout_fd, "r");
+            var stdout_text = new char[1024];
+            f.gets (stdout_text);
+            id = int.parse ((string)stdout_text);
+        } catch (Error e) {
+            warning (e.message);
```
Comment (jeremypw) [inline src/SessionWidget.vala:194]: Is it right to continue constructing the window if something when wrong with spawning the process?

File: src/SessionWidget.vala [hunk 2]
```diff
@@ -107,6 +140,82 @@ public class A11Y.SessionWidget : Gtk.Grid {
...
+                string[] argv;
+                Shell.parse_argv ("orca --replace", out argv);
+                Process.spawn_async (null, argv, null, SpawnFlags.SEARCH_PATH, null, out reader_pid);
+            } catch (Error e) {
+                warning (e.message);
+            }
+        } else {
+            Posix.kill (reader_pid, Posix.Signal.KILL);
+            Posix.waitpid (reader_pid, out status, 0);
+            reader_pid = 0;
```
Comment (jeremypw) [inline src/SessionWidget.vala:169]: This repeats some code in the destructor

File: src/SessionWidget.vala [hunk 3]
```diff
@@ -107,6 +140,82 @@ public class A11Y.SessionWidget : Gtk.Grid {
...
+            id = int.parse ((string)stdout_text);
+        } catch (Error e) {
+            warning (e.message);
+        }
+
+        var keyboard_socket = new Gtk.Socket ();
+        keyboard_window = new Gtk.Window ();
+        keyboard_window.type_hint = Gdk.WindowTypeHint.DOCK;
+        keyboard_window.accept_focus = false;
+        keyboard_window.focus_on_map = false;
```
Comment (jeremypw) [inline src/SessionWidget.vala:201]: Does not use recommended code format for object construction.

File: src/SessionWidget.vala [hunk 4]
```diff
@@ -18,11 +18,23 @@
  */
 
 public class A11Y.SessionWidget : Gtk.Grid {
+    public Wingpanel.IndicatorManager.ServerType server_type { get; construct; }
+
+    private KeyFile settings;
```
Comment (jeremypw) [inline src/SessionWidget.vala:23]: I would name this to indicate it is only used in a greeter session.

File: src/SessionWidget.vala [hunk 5]
```diff
@@ -107,6 +140,82 @@ public class A11Y.SessionWidget : Gtk.Grid {
...
+        }
+
+        if (reader_pid != 0) {
+            Posix.kill (reader_pid, Posix.Signal.KILL);
+            Posix.waitpid (reader_pid, out status, 0);
+            reader_pid = 0;
+        }
+    }
+
+    private void toggle_screen_reader (bool active) {
```
Comment (jeremypw) [inline src/SessionWidget.vala:157]: I would name (or comment) this to indicate it is only called in a greeter session.  May add `requires settings != null`.

File: src/SessionWidget.vala [hunk 6]
```diff
@@ -55,27 +67,52 @@ public class A11Y.SessionWidget : Gtk.Grid {
...
+                toggle_screen_reader (screen_reader.active);
+            });
+
+            onscreen_keyboard.notify["active"].connect (() => {
+                toggle_keyboard (onscreen_keyboard.active);
+            });
+
             try {
-                AppInfo.launch_default_for_uri ("settings://universal-access", null);
+                onscreen_keyboard.active = settings.get_boolean ("greeter", "onscreen-keyboard");
```
Comment (jeremypw) [inline src/SessionWidget.vala:111]: What is the point of this? You just set that setting to "false" a few lines back.  Can't you just set the keyboard switch to `off` directly, with out a try/catch clause?

**Review by jeremypw — CHANGES_REQUESTED** [review]
Added some comments after looking at code - have not tried it out yet.

Tbh, I cannot immediately see why you need the `GLib.KeyFile` at all in the greeter session. It does not seem to be used for anything (?)

**Comment by danirabbit** [comment]
Hey sorry to waste your time on this @jeremypw. This should probably have been closed a long time ago or at least set as draft!

**Pull request closed without merging**